### ISSUE TITLE
Is there any way I can draw and animate the radius of a circle from its center?

### ISSUE THREAD

**Issue by 9itish** [craftyjs/Crafty#1055]
I am using the following code to create a circle:

```
Crafty.c("Circle", {
            init: function (radius, color) {
                this.requires("2D, Canvas");
                this.bind("Draw", this._draw_me);
                this.ready = true;
            }
            , _draw_me: function (e) {
                var ctx = e.ctx;
                ctx.beginPath();
                ctx.fillStyle = this.color;
                ctx.arc(e.pos._x + this.radius, e.pos._y + this.radius, this.radius, 0, Math.PI * 2);
                ctx.closePath();
                ctx.fill();
            }
});
```

This draws my circle using the  top left corner as the pivot point. On using the following line to draw the arc, the pivot point changes to center which is what I want.

```
 ctx.arc(e.pos._x, e.pos._y, this.radius, 0, Math.PI * 2);
```

I also have to animate the radius of the circle that I am drawing. So, I am using the following code:

```
 .bind("EnterFrame", function (eventData) {
            if(this.radius < 0.03*gHeight && this.increase) {
                this.w += 1;
                this.h += 1;
                this.radius += 0.5;
            }
           More such code...
```

If I use the first version of the circle, then the circle seems to increase in size from the top left corner. If I use the second version the circle increases in size from the center but the circle is not drawn properly. Only the bottom right quadrant of the circle animates.

Here is a fiddle because I am sure it was hard to follow what I was saying : https://jsfiddle.net/fn0etaxq/

Is there some way I can make the circle change size from the center and still be animated properly?

**Comment by 200sc** [comment]
The problem is attempting to grow around the x,y when the x,y is the top left corner-- Crafty defines the areas that need to be redrawn by the rectangle from x to x+w, y to y+h for each entity. So when you try to draw above or to the left of the x,y values you don't actually get to draw onto a part of the canvas that is being redrawn. 

https://jsfiddle.net/fn0etaxq/2/

The above fiddle moves the center to the middle of the drawn rectangle. That's not what you want, but you'll need logic that starts there at the very least. I would recommend not changing width and height.

**Comment by 9itish** [comment]
Hey @Sythe2o0, thanks for the fiddle. :)

As far as I can tell, using `ctx.arc((this.x+this.w/2)+1, (this.y+this.h/2)+1, this.radius-1, 0, Math.PI * 2)` is tha same thing as using `ctx.arc(this.x + this.radius, this.y + this.radius, this.radius, 0, Math.PI * 2);`. Am I missing something?

**Comment by 200sc** [comment]
I wrote the code very hastily, those statements look equivalent but I think I was adding in the minus one because it was going over the width and height limits sometimes. In general, I've found it very hard to work with any sort of drawing direct onto a context in Crafty and it often has huge performance impacts.

**Comment by 9itish** [comment]
I forgot to mention that if you go to the demo (https://jsfiddle.net/fn0etaxq/) and press arrow keys the black circle will start moving which is expected. However, the smaller red circle starts drawing itself properly, I mean it increases in size from its center and redraws the whole circle. Why is that?

**Comment by 200sc** [comment]
Because when the other entity starts to move, the entire screen is marked to be redrawn. You can go into some settings on chrome, for example, to see what part of the screen is being redrawn at any given point in time. F12 -> The menu to the left of 'Console' -> Rendering -> check on 'Paint Flashing'

**Comment by 9itish** [comment]
Thanks for the clarification.  :)
Can I force the entire screen to be redrawn from the beginning? I will only have thirteen or fourteen entities at most at the screen at once. Will it still cause performance issues?

**Comment by 200sc** [comment]
I don't recall if there's a way to do it in Crafty that's easy, I remember doing some tricks to force redraws but I don't remember what they were, sorry. Something like having an invisible entity shift back and forth might do it.

**Comment by 9itish** [comment]
Just one more question, actually two. :)

Is there any flag that I can turn no to see the bounding boxes of all entities for easy debugging?

How can I bind `EnterFrame` to entities generated later at random intervals? This is my code right now:

```
// Create one entity in the beginning and then rest of the entities randomly

setInterval(function() {
    if (Math.random() < 0.3) {
        Crafty.e("Eatables, 2D, Canvas, Color, Circle, Collision, Motion").attr({
                x: getRandomInt(0,gWidth/2)
                , y: getRandomInt(0,gHeight/2)
                , w: 0.04*gHeight
                , h: 0.04*gHeight
                , radius: 0.02*gHeight
                , color: 'brown'
                , increase: true
        });
    }
},2000);

Crafty("Eatables").bind("EnterFrame", function (eventData) {
            if(this.radius < 0.022*gHeight && this.increase) {
                this.w += 0.5;
                this.h += 0.5;
                this.radius += 0.25;
            }

            More code
 });
```

The problem here is that only the first entity created manually is animated and not the rest of them. How can I animate all of them?

**Comment by 200sc** [comment]
The closest thing to the highlight you want is WiredHitBox or SolidHitBox, which each highlight collision areas of their entities. 

As for binding to random entities, you should have the binding be a part of some component that the entities implement, or if you need some local data from when you create it, have it bind in a function call after storing the result of Crafty.e().
